Source/items.cpp

```diff
@@ -3292,6 +3292,75 @@ Item *SpawnUnique(_unique_items uid, Point position, std::optional<int> level /*
 	return &item;
 }
 
+void TryRandomUniqueItem(Item &item, _item_indexes idx, int8_t mLevel, int uper, bool onlygood, bool pregen)
+{
+	// If the item is a non-quest unique, find a random valid uid and force generate items to get an item with that uid.
+	if ((item._iCreateInfo & CF_UNIQUE) != 0 && item._iMiscId != IMISC_UNIQUE) {
+		int8_t itemLvl = mLevel;
+
+		// mLevel remains unadjusted when arriving in this function, and there is no call to set iblvl until CheckUnique(), so we adjust here.
+		if (uper == 15)
+			itemLvl += 4;
+
+		std::vector<std::pair<int, int>> uids; // Contains uid and uidOffset.
+
+		// Gather all potential unique items. uid is the index into UniqueItems.
+		for (int uid = 0; uid < static_cast<int>(UniqueItems.size()); ++uid) {
+			const auto &uniqueItem = UniqueItems[uid];
+
+			// Verify the unique item base item matches idx.
+			bool isMatchingItemId = uniqueItem.UIItemId == AllItemsList[static_cast<size_t>(idx)].iItemId;
+			// Verify itemLvl is at least the unique's minimum required level.
+			bool meetsLevelRequirement = itemLvl >= uniqueItem.UIMinLvl;
+			// Verify item hasn't been dropped yet. We set this to true in MP, since uniques previously dropping shouldn't prevent further identical uniques from dropping.
+			bool uniqueNotDroppedAlready = !UniqueItemFlags[uid] || gbIsMultiplayer;
+
+			if (IsUniqueAvailable(uid) && isMatchingItemId && meetsLevelRequirement && uniqueNotDroppedAlready)
+				uids.emplace_back(std::make_pair(uid, 0));
+		}
+
+		// Set all uidOffset values for each entry in uids vector.
+		for (int i = static_cast<int>(uids.size() - 1); i >= 0; --i) {
+			int currentUidOffset = 0;
+			for (int j = i + 1; j < uids.size(); ++j) {
+				if (UniqueItems[uids[i].first].UIItemId == UniqueItems[uids[j].first].UIItemId && UniqueItems[uids[i].first].UIMinLvl == UniqueItems[uids[j].first].UIMinLvl) {
+					currentUidOffset = std::max(currentUidOffset, uids[j].second + 1);
+				}
+			}
+			uids[i].second = currentUidOffset;
+		}

```

Comment:
Since this nested loop exclusively operates on the `uids` collection, and the `uids` collection was filtered based on `UniqueItemFlags`, this will get a wrong answer considering how the `CheckUnique()` function uses this during `RecreateItem()` calls.

Reply:
This function is only called when generating new items. It has no interaction with item recreation.

Reply:
Yeah, it doesn't call item recreation directly. But it does compute a uid offset which is encoded into `dwBuff` and then used during item recreation.

Reply:
Oh I see, if items are already found in SP, we can end up with the incorrect uid offset when recreating

Reply:
Okay, I fixed it by not omitting already found uniques from `uids` vector, and then when randomly selecting a uid from the vector, we skip over ones that are already found. If we exceed a counter when searching for a unique, then we force generate a non-unique item, as it would in vanilla.

Reply:
Sounds more complicated than what I would have done. Food for thought:

```c++
// Set all uidOffset values for each entry in uids vector.
for (auto &pair : uids) {
	for (int j = pair.first + 1; j < UniqueItems.size(); ++j) {
		if (UniqueItems[pair.first].UIItemId == UniqueItems[j].UIItemId && UniqueItems[pair.first].UIMinLvl == UniqueItems[j].UIMinLvl) {
			pair.second++;
		}
	}
}
```